File: interface/src/events.ts
/*

Copyright (c) 2026-present, RoRvzzz. All rights reserved.

https://rorvzzz.cool

*/
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export type EventName =
  | "bootstrap_status"
  | "progress"
  | "prompt_required"
  | "connectivity_error"
  | "fatal_error"
  | "watcher_activity";

export interface EventPayload {
  timestamp: string;
  event: EventName;
  payload: unknown;
}

export async function subscribeToCoreEvents(
  onEvent: (payload: EventPayload) => void
): Promise<UnlistenFn[]> {
  const eventNames: EventName[] = [
    "bootstrap_status",
    "progress",
    "prompt_required",
    "connectivity_error",
    "fatal_error",
    "watcher_activity"
  ];

  const unlistenFns = await Promise.all(
    eventNames.map((eventName) =>
      listen<unknown>(eventName, (event) => {
        onEvent({
          timestamp: new Date().toISOString(),
          event: eventName,
          payload: event.payload
        });
      })
    )
  );

  return unlistenFns;
}

File: core/src/events.rs
/*

Copyright (c) 2026-present, RoRvzzz. All rights reserved.

https://rorvzzz.cool

*/
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PromptKind {
    ConfirmLaunch,
    ChannelChange,
    UpdateAvailable,
    InstallLocationMismatch,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case", tag = "event")]
pub enum WatcherEvent {
    Started,
    Stopped,
    ActivityChanged { description: String },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "event", rename_all = "snake_case")]
pub enum DomainEvent {
    BootstrapStatus { message: String },
    Progress { current: u64, total: u64 },
    PromptRequired { kind: PromptKind, message: String },
    ConnectivityError { title: String, description: String },
    FatalError { code: String, message: String },
    WatcherActivity { activity: WatcherEvent },
}
